add filter using a raster method, remove flipud! from reduce

Crghilardi opened the pull request:
Closes #5 

This PR add a `Base.filter!(index::PointCloudRasterizers.PointCloudIndex, raster::GeoArray, condition=nothing)` method. 
I also removed the `flipud!` which was preventing this from working correctly. I also updated the README and tests.

Crghilardi committed:
add filter using a raster method, remove flipud from reduce

visr commented:
Nice, thanks for adding this method!

Don't we get an upside down reduced raster now on `reduce` without `flipud`? I was thinking that we would need to modify the index itself, such that we don't need `flipud` anymore.

Also, I know that the existing tests only just run the code and not actually test anything, but it would be great to move away from that. Even just a simple count or mean of the reduced results can prevent regressions in the future.

Crghilardi commented:
I don't quite understand what the `flipud!` was doing in the first place, so I don't know if I am comfortable changing the indexing myself.

I did notice the testset was a copy paste from the README but wasn't sure how important that was. I can do some work on that and wrap everything in `testsets`. Would you want the test changes in this branch or a new one?

Gelukkig Nieuwjaar

evetion commented:
`flipud!` reverses the `GeoArray` in the y direction and flips the y step in the `AffineMap` as well. If you would open both `.tif` files in QGIS, you would see no difference. However, software packages often have a preference for y coordinates increasing (Julia/Python), or decreasing (GDAL). `flipud` can help with that.

Crghilardi committed:
begin to refactor tests

Crghilardi commented:
I added a commit that starts to restructure the tests. It is failing on CI but `julia .test/runtests.jl`passes for me locally? 

I can try to do the index work if you can describe exactly what needs to be done. I don't understand the `reduce` function clearly to make big changes.

Crghilardi committed:
correct test values

visr commented:
I had a chat about this with @evetion but he basically said since the y axis direction varies between different packages, it's hard to 'get it right' in the index. So let's leave the index as it is. We can flip it as needed when we are forced to, for instance when saving it with GDAL.

Thanks for this, it looks good to me, I'll merge it now.

evetion commented on README.md:
```diff
@@ -31,7 +31,7 @@ pointcloud = LazIO.open(lazfn)
 raster_index = index(pointcloud, cellsizes)
 
-#get some information about the index
+##get some information about the index:
```
Double # and the : is not required

evetion commented on README.md:
```diff
@@ -54,6 +54,14 @@ filter!(raster_index, last_return)
 Filters are done in-place and create a new index matching the condition. It does not change the loaded dataset.
 
+Filtering can also be done compared to a computed surface.
+For example, if we want to select all points within some tolerance of the median raster from above:
```
The raster is below, we have to think how to reword/move it to have an understandable flow.

evetion commented on src/PointCloudRasterizers.jl:
```diff
@@ -74,6 +74,28 @@ function Base.filter!(index::PointCloudIndex, condition=nothing)
+"""Filter out index on given condition relative to a reduced layer."""
+function Base.filter!(index::PointCloudRasterizers.PointCloudIndex, raster::GeoArray, condition=nothing)
+
+    #check that size and affine info match
```
I will move this check to GeoArrays eventually, good to have an equal operation for all fields/sizes except for the values in the array.

evetion commented on test/runtests.jl:
```diff
@@ -1,29 +1,30 @@
+raster = reduce(idx, field=:Z, reducer=median)
 
-# Filter on last returns (inclusive)
+#indexing and reducing tests
```
This is nitpicking, but I do like comments with an extra space like `# so`.

evetion reviewed (changes requested):
Thanks for taking the time to make a PR! Some comments, mostly nitpicks apart from the explanation in the README.

evetion commented:
The `flipud!` should be removed from this one. I think the positive x, y flows naturally from coding it and will work in all dimensions. Suddenly flipping only the y seems weird. Both GeoStats and Plots need positive axes, it's mostly GDAL, when finally serialized to tiff, complains about it.

Feel free to rework things into `@testset`s in this branch.

Crghilardi committed:
make review changes:  add spaces, modify orfer of README

Crghilardi commented:
OK, I addressed your comments in the latest commit. I re-ordered the README so filtering comes after `raster` is declared and cleaned up spacing on various comments. `flipud!` remains fully removed. 

I originally wrote the tests in `@testsets` but ran into scoping issues. I can revisit tests at a later time, I would like to get this functionality merged so I can finish writing my blogpost.
